### PR TITLE
DO NOT MERGE [wip] starting the porting to semantic math

### PR DESCRIPTION
**Pull request by evykassirer**
this is a complete work in progress, but here's what I got done tonight

The basic idea was to replace the bodies of Node.Type and Node.Creator with their semantic-math equivalents, and see if the rest of the code would just work. I've found a bunch of things that need updating in this repo or in semantic-math repos and added TODOs, and will keep investigating why various things are failing to find more TODOs.

This isn't linted and the commit names aren't very descriptive, so I'll try to get that done. If you see something here and want to fix it in semantic-math or help me fix something please yes help! :)

My goal is to in the next week (or so?) get a better idea of what we need to do, clean up this PR, and then make this a base branch with a bunch of TODOs and disabled tests, and people can make PRs off of it to help fix various issues. (Let me know if you think that won't work)

WOOOO porting :)

cc @kevinbarabash 

More things, not mentioned in this PR's code:
- [ ] when printing in math-parser I got `throw new Error('unrecognized node');` -- can it show what the node looks like? easier to debug that
- [ ] we should make sure we don't need flatten and then get rid of it
- [ ] once this is stable, we should look into getting rid of paren nodes when not needed now that math-parser handles printing them by tree structure

### CODE REVIEW

File: lib/Symbols.js [hunk 1]
```diff
@@ -12,7 +13,7 @@ Symbols.getSymbolsInEquation = function(equation) {
 
 // return the set of symbols in the expression tree
 Symbols.getSymbolsInExpression = function(expression) {
-  const symbolNodes = expression.filter(node => node.isSymbolNode); // all the symbol nodes
+  const symbolNodes = filter(expression, Node.Type.isSymbol); // all the symbol nodes
```
Comment (evykassirer) [inline lib/Symbols.js]: this should be called `findAll` or something more descriptive - and maybe it'd be better living in semantic-math as a function on the node?

**Comment by kevinbarabash** [comment]
I can definitely add better debug messages to the `math-parser`.  Please open issues in that repo.

**Comment by evykassirer** [comment]
I got to 

```
 426 passing (4s)
  35 pending
  221 failing
```

pending means I know why it's failing and don't want to fix it in this PR (for the most part)
the failing ones I haven't looked into all of them, but I think at least half are more printing things

one is that `nthroot`s need to be updated to explicitly show that the root is 2 (which I think is fine to change on mathsteps side)

another is to get rid of the space between coefficient and symbol when the symbol is also of the form `x^2`, so I'll make a PR to math-parser for that this weekend hopefully

File: lib/node/Creator.js [hunk 2]
```diff
@@ -77,8 +76,7 @@ const NodeCreator = {
 
   // Given a root value and a radicand (what is under the radical)
   nthRoot (radicandNode, rootNode) {
-    const symbol = NodeCreator.symbol('nthRoot');
-    return new math.expression.node.FunctionNode(symbol, [radicandNode, rootNode]);
+    return build.nthRoot(radicandNode, rootNode); // TODO: rename this to index everywhere? or rename index in math-nodes?
```
Comment (evykassirer) [inline lib/node/Creator.js]: ah ok yeah it should be http://www.purplemath.com/modules/radicals.htm

**Comment by evykassirer** [comment]
eeeek it's getting close! just a few more things to clean up!

my plans are:

1. get all the tests passing or skipped
2. return to skipped tests and decide if I want to deal with them in this PR or leave as a TODO
3. address all the semantic math and porting related TODOs
4. go through all these changes and self-review and cleanup while cleaning them into more reviewable commits
5. see if folks from the community will review this for me?? (fingers crossed)
6. merge the new tree!!

**Comment by evykassirer** [comment]
down to 10 failing tests :o

**Comment by evykassirer** [comment]
600 passing (1s)
  82 pending

and the linter passes!

🎉 

next steps:
- return to skipped tests and decide if I want to deal with them in this PR or leave as a TODO
- go through the semantic-math related TODOs - some of them @kevinbarabash has already fixed and I can update already, some are left to be done
- go through everything myself to see if it's in an okay state for other folks to review, do some cleanup on rough edges
- ask around for anyone interested in taking a look - I can also go through and put the code into cleaner isolated commits if that helps! but that'll take some work so I'll only do it if it makes things a lot easier to review. I know Ahmed already said he could take a look, hopefully Kevin is willing as well!

**Pull request closed without merging**